### DISCUSSION TITLE
Adding Podman Support to torchx

### DISCUSSION THREAD

**Discussion by KPostOffice** [containers/podman-py#258] [Q&A] [answered]
Hi all, I'm looking into what it would take to add podman support to torchx. This library seemed like it would be a good fit for a drop in replacement for the Docker client. I changed the code to use a podman client if it is present on the users system. However, when testing the changes I ran into errors with the file object that is created on [this](https://github.com/pytorch/torchx/blob/65ed0fd2b4be278e396b055625050a1a1cd32bda/torchx/workspace/docker_workspace.py#L236) line. When I pass the resulting fileobj to the build function I get the following error:
```
  File "python3.8/site-packages/podman/domain/images_build.py", line 88, in build
    shutil.copyfileobj(kwargs["fileobj"], file)
  File "/usr/lib64/python3.8/shutil.py", line 208, in copyfileobj
    fdst_write(buf)
TypeError: write() argument must be str, not bytes
```
I tried changing the mode of the file that is being copied to to be `wb+` as a quick workaround, however it resulted in another error:
```
podman.errors.exceptions.BuildError: building at STEP " ": Build error: Unknown instruction: "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x
...
&imagebuilder.Step{Env:[]string{"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"}, Command:"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\
...
```
Any suggestions on how to proceed would be much appreciated.

**Comment by rhatdan** [comment]
@umohnani8 @jwhonce PTAL

**Comment by KPostOffice** [comment, marked as answer]
I didn't realize that the Podman REST api was compatible with the Docker REST api. I was able to solve my use case by manually setting `DOCKER_HOST` after running `$ podman system service -t 0 tcp://localhost:8889`
